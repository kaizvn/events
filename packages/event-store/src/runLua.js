/* @flow */
import promisify from 'es6-promisify';
import crypto from 'crypto';

type EvalParams = {
  keys: ?(string[]),
  argv: ?(string[]),
};

type RunLua = (any, string, EvalParams) => Promise<any>;

const sha1 = (text: string): string => {
  const hash = crypto.createHash('sha1');
  return hash.update(text).digest('hex');
};

const map = new Map();

const runLua: RunLua = promisify(
  (redisClient, luaScript: string, { keys, argv }: EvalParams, cb) => {
    const nonEmptyKeys = keys || [];
    const nonEmptyArgv = argv || [];

    const hashOrNull = map.get(luaScript);
    if (hashOrNull) {
      return redisClient.evalsha(
        hashOrNull,
        nonEmptyKeys.length,
        ...nonEmptyKeys,
        ...nonEmptyArgv,
        cb
      );
    }

    return redisClient.eval(
      luaScript,
      nonEmptyKeys.length,
      ...nonEmptyKeys,
      ...nonEmptyArgv,
      (err, val) => {
        if (!err) {
          const hash = sha1(luaScript);
          map.set(luaScript, hash);
        }

        cb(err, val);
      }
    );
  }
);

export default runLua;
