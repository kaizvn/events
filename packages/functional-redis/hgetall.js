const promisify = require('es6-promisify');
const curry = require('curry');

module.exports = promisify((key, client, cb) => client.hgetall(key, cb));
