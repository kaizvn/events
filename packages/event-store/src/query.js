/* @flow */
import { createClient, RedisClientType } from './redis-client';

import type { CommitConfig } from '../types/Config.type';
import runLua from './runLua';

const queryLua = `
local function getKeys(from, to, key)
  local newArray = {};
  for id=from,to do
    table.insert(newArray, {id, redis.call('hget', key, id)});
  end

  return newArray
end

local from = tonumber(ARGV[1]);
local to = tonumber(ARGV[2]) or redis.call('HLEN', KEYS[1]);

return getKeys(from, to, KEYS[1])
`;

export const query = async (client: RedisClientType, namespc: string, ...argv: number[]) => {
  const array = await runLua(client, queryLua, {
    keys: [`${namespc}::events`],
    argv: argv.map(String),
  });

  return array.map(([id, event]) => ({
    id,
    ...JSON.parse(event),
  }));
};

export default (config: CommitConfig) => {
  const client = createClient(config.redis);

  return async (req: any) => {
    const lastEventId = Number(req.headers['Last-Event-ID'] || req.query.lastEventId);
    const events = await query(client, config.namespc, lastEventId);

    // we trust the input from commit. otherwise we have to do a try-parse
    return events;
  };
};
