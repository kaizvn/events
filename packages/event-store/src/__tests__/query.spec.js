/* @flow */
import flushdb from 'redis-functional/flushdb';
import hgetall from 'redis-functional/hgetall';

import query from '../query';
import redisClient from '../redis-client';
import runLua from '../runLua';

describe('query endpoint', () => {
  it('should query', async () => {
    await flushdb(redisClient);

    await runLua(
      redisClient,
      `
        for id=1,10 do
          redis.call('HSET', 'test::events', id, '{"type": "test", "payload":'..id..'}')
        end
      `,
      { argv: [], keys: [] }
    )

    const req = {
      query: {
        lastEventId: 5
      },
      headers: {}
    };

    const service = query({ namespc: 'test' });

    const actual = await service(req);
    expect(actual).toMatchSnapshot();
  });
});
