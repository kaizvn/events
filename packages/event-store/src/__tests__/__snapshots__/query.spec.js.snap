// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query endpoint should query 1`] = `
Array [
  Object {
    "payload": 5,
    "type": "test",
  },
  Object {
    "payload": 6,
    "type": "test",
  },
  Object {
    "payload": 7,
    "type": "test",
  },
  Object {
    "payload": 8,
    "type": "test",
  },
  Object {
    "payload": 9,
    "type": "test",
  },
  Object {
    "payload": 10,
    "type": "test",
  },
]
`;
